oeps/oep-0037-bp-test-data.rst

```diff
@@ -0,0 +1,254 @@
+========
+
+Proposes best practices in creating and maintaining sample data for use in
+Open edX test cases and development environments.
+
+Motivation
+==========
+
+When testing or developing an Open edX service, it us usually necessary
+to have some data already loaded and configured: users, courses, grades, etc.
+This is true for unit tests, browser automation tests, load tests,
+development work in devstack, and more.  There are many different ways of
+creating and loading such data, and in the absence of concrete guidance on
+best practices, many of them have been (inconsistently) used in Open edX.
+This makes it difficult to decide how to prepare such data, to learn how to
+load it when needed, and also to keep it working correctly as the code
+changes over time.
+
+This OEP attempts to provide the guidance needed to manage this data
+logically and consistently in order to streamline development, refactoring,
+and automated testing.
+
+Specification
+=============
+
+There are a few different aspects of test data management, each of which
+has different tools that can make it more painless:
+
+Reusing Data
+------------
+
+Once a development or production environment has been populated with a data
+set comprehensive enough to exercise certain functionality in Open edX, it can
+be useful to "save it" for later use.  The first step is to make sure that any
+`PII`_ has been stripped from the data.  This is a particular concern for
+production data that may include customer information, but even development
+data may include a developer's email address and other personal information.
+Tools such as `dj_anonymizer`_ can be useful in making a first pass at
+stripping such PII, but it should be followed by manual verification against
+the `OEP-30`_ annotations.
+
+Once PII has been removed, an attempt can be made to export just the needed
+data.  This may be a combination of course export to OLX and generation of
+fixtures from selected relational data using a utility such as
+`django-fixture-magic`_.  These fixtures should not be saved verbatim, but
+rather used as a guide for producing Python code that generates roughly
+equivalent data using ``factory_boy`` as described below.
+
+Note that if the following recommendations on creating data are followed, test
+data created for a unit test can be fairly easily re-purposed for use as
+development environment data sets.  This can be useful because the starting
+point is already free of PII and has been verified by tests to work as
+intended.
+
+.. _PII: https://open-edx-proposals.readthedocs.io/en/latest/oep-0030-arch-pii-markup-and-auditing.html
+.. _dj_anonymizer: https://github.com/preply/dj_anonymizer
+.. _OEP-30: https://open-edx-proposals.readthedocs.io/en/latest/oep-0030-arch-pii-markup-and-auditing.html
+.. _django-fixture-magic: https://github.com/davedash/django-fixture-magic
+
+Creating Data
+-------------
+
+Whether defining it from scratch or basing it on a dump of data from an
```

Comment:
@jmbowman Why don't we load those data seamlessly without having to worry about writing the commands and integrating it in all backends? I was suggesting here having some sort of test case that loads the data for every test case it hits in certain path dynamically. So for example if I'm running:

```
./manage.py test ecommerce.courses.tests.test_utils
```
The runner loads data from `ecommerce.courses.tests.data.*` seamlessly. We still can use the `load_test_data` command to perform this operation, but what do you think?

Reply:
Currently, the test data required for each test case can vary widely even between different methods in the same `TestCase` subclass.  Instead of guessing at what data is needed for a given test case based on where it is, I'm thinking we'd be better off splitting the data setup out into separate functions which can be shared by multiple test cases (even in different areas of the code) which call them during `setUpTestData`, `setUp`, or the test method itself as appropriate.